src/Marketing/InstalledExtensions.php

```diff
@@ -152,6 +158,38 @@ protected static function get_facebook_extension_data() {
 		return $data;
 	}
 
+	/**
+	 * Get Pinterest extension data.
+	 *
+	 * @return array|bool
+	 */
+	protected static function get_pinterest_extension_data() {
```

Comment:
This is just an optional idea, especially as I know you've followed the same format the other plugins have used in this plugin. Feel free to ignore based on your priorities 🙂 

We have a goal to create more unit tests, however `get_pinterest_extension_data` looks difficult to create unit tests for because it requires the Pinterest plugin to be installed.

A possible alternative is to use filters that are added by the Pinterest plugin. For example, `PluginsHelper::is_plugin_installed( $slug )` could be replaced by:

```
apply_filters( 'wc_admin_is_pinterest_plugin_installed', false )
```

And in the Pinterest plugin, a filter could be added to notify WC Admin about the state:

```
function is_installed( $installed ) {
    return true;
}
add_filter( 'wc_admin_is_pinterest_plugin_installed', 'is_installed', 10, 1 );
```

Likewise, a similar technique could be used for `Pinterest_For_Woocommerce()::is_setup_complete()` 

This would then allow unit tests to be written, as the test themselves can use these filters to mock the different states of the Pinterest plugin. A downside is that some other plugin might mess with things, but they could do that anyway with WordPress being as open as it is.

Reply:
> We have a goal to create more unit tests, however `get_pinterest_extension_data` looks difficult to create unit tests for because it requires the Pinterest plugin to be installed.

Sounds like an integration test, with Pinterest installed, would allow us to test the behaviour that could break this API (and this screen). I think this is somewhat consistent with WordPress PHP tests - they are sometimes (often?) used [more like integration tests](https://make.wordpress.org/cli/handbook/misc/plugin-unit-tests/) than pure unit tests. 

If we added the hooks and then unit-tested this class in isolation, what bugs might that catch? It feels like we'd be testing implementation details rather than meaningful, robust behaviour.

We could add a unit test that confirms that each `get_*_extension_data` returns an object with the expected shape (e.g. fields and types, perhaps confirm URLs are valid). But this would of course only test the code path where the extension(s) are not installed, which is less critical – having a test that confirms it doesn't crash (with no extension present) is probably all that's needed.

Tests are tricky! I try to avoid adding tests of implementation details, and focus on testing core behaviour, though this is complex in WordPress as self-contained (no dependencies) code is rare.

Reply:
Note – when Pinterest for WooCommerce is installed it might be possible to install it as part of a test suite! We could trial this with one of the other extensions first and see if it helps confirm this module's behaviour via tests 🍇

Reply:
Thanks @haszari! I totally agree tests are tricky!

> If we added the hooks and then unit-tested this class in isolation, what bugs might that catch? It feels like we'd be testing implementation details rather than meaningful, robust behaviour.

The test case originally coming to mind was checking the `utm=` parameter is in the docs URL. It’s now a product requirement of WC Admin for outgoing links to woocommerce.com so it would be good to protect it from being lost in future changes.

I also think it would also be good to have the logic behind `settingsUrl` tested.

> Sounds like an integration test, with Pinterest installed, would allow us to test the behaviour that could break this API (and this screen). I think this is somewhat consistent with WordPress PHP tests - they are sometimes (often?) used more like integration tests than pure unit tests.

I do agree WordPress PHP unit tests are more like integration tests than pure units. That said while I think it's a nice idea, I suspect it would be too “heavy” performance-wise installing the plugin/extension in the tests. I think it might also also be tricky for the community to run tests when extensions require a subscription to install them. I think I’d prefer to be able to mock the plugin’s state in some way.

A different approach to filters might just be with some better OOP techniques on the `InstalledExtensions` class. For example adding some basic methods to the class that the plugins/extensions would use to notify the class about their state

I'm glad we had the discussion! I’m happy to pause this for now but I think it would be great to come back to again some time 👍